### ISSUE TITLE
Heavy disk usage on LXD 4 start, compared with LXD 3

### ISSUE THREAD

**Issue by marius-nicolae** [lxc/lxd#8748]
# Required information

 * Distribution: Ubuntu
 * Distribution version: 20.04
 * The output of "lxc info" or if that fails:
   * Kernel version: 5.4.0-72-generic
   * LXC version:  3.0.4 and 4.0.6 or 4.13
   * LXD version: 3.0.4 and 4.0.6 or 4.13
   * Storage backend in use: zfs

# Issue description

For a clean LXD install (no containers), I noticed that the hard drive (a WD black) is continuously used for approx 2 - 3 minutes on LXD 4 start (the disk activity continues even after "lxd.daemon[10046]: => LXD is ready" log message) while for LXD 3, on the same machine, the disk activity stops after a few seconds (up to 10 seconds). The behavior is noticed at every "systemctl restart snap.lxd.daemon" and it manifests even without "lxd init".

There is nothing relevant in the logs which could pinpoint the problem.

# Steps to reproduce

For LXD 3 the reproducing steps are:

 1. Install LXD : "snap install lxd --channel=3.0/latest"
 2. systemctl restart snap.lxd.daemon
 3. The disk activity should stop after at most 10 seconds.

For LXD 4 the reproducing steps are:

 1. Install LXD : "snap install lxd --channel=4.0/latest"
 2. systemctl restart snap.lxd.daemon
 3. The disk activity can be notices for approx 2 - 3 minutes, way longer than LXD 3.

# Information to attach

 - [ ] Main daemon log (at /var/log/lxd/lxd.log or /var/snap/lxd/common/lxd/logs/lxd.log)
8<
t=2021-05-06T16:16:11+0300 lvl=info msg="LXD 4.0.6 is starting in normal mode" path=/var/snap/lxd/common/lxd
t=2021-05-06T16:16:11+0300 lvl=info msg="Kernel uid/gid map:"
t=2021-05-06T16:16:11+0300 lvl=info msg=" - u 0 0 4294967295"
t=2021-05-06T16:16:11+0300 lvl=info msg=" - g 0 0 4294967295"
t=2021-05-06T16:16:11+0300 lvl=info msg="Configured LXD uid/gid map:"
t=2021-05-06T16:16:11+0300 lvl=info msg=" - u 0 1000000 1000000000"
t=2021-05-06T16:16:11+0300 lvl=info msg=" - g 0 1000000 1000000000"
t=2021-05-06T16:16:11+0300 lvl=info msg="Kernel features:"
t=2021-05-06T16:16:11+0300 lvl=info msg=" - closing multiple file descriptors efficiently: no"
t=2021-05-06T16:16:11+0300 lvl=info msg=" - netnsid-based network retrieval: yes"
t=2021-05-06T16:16:11+0300 lvl=info msg=" - pidfds: yes"
t=2021-05-06T16:16:11+0300 lvl=info msg=" - uevent injection: yes"
t=2021-05-06T16:16:11+0300 lvl=info msg=" - seccomp listener: yes"
t=2021-05-06T16:16:11+0300 lvl=info msg=" - seccomp listener continue syscalls: yes"
t=2021-05-06T16:16:11+0300 lvl=info msg=" - seccomp listener add file descriptors: no"
t=2021-05-06T16:16:11+0300 lvl=info msg=" - attach to namespaces via pidfds: no"
t=2021-05-06T16:16:11+0300 lvl=info msg=" - safe native terminal allocation : yes"
t=2021-05-06T16:16:11+0300 lvl=info msg=" - unprivileged file capabilities: yes"
t=2021-05-06T16:16:11+0300 lvl=info msg=" - cgroup layout: hybrid"
t=2021-05-06T16:16:11+0300 lvl=warn msg=" - Couldn't find the CGroup blkio.weight, disk priority will be ignored"
t=2021-05-06T16:16:11+0300 lvl=warn msg=" - Couldn't find the CGroup memory swap accounting, swap limits will be ignored"
t=2021-05-06T16:16:11+0300 lvl=info msg=" - shiftfs support: disabled"
t=2021-05-06T16:16:11+0300 lvl=info msg="Initializing local database"
t=2021-05-06T16:16:11+0300 lvl=info msg="Starting /dev/lxd handler:"
t=2021-05-06T16:16:11+0300 lvl=info msg=" - binding devlxd socket" socket=/var/snap/lxd/common/lxd/devlxd/sock
t=2021-05-06T16:16:11+0300 lvl=info msg="REST API daemon:"
t=2021-05-06T16:16:11+0300 lvl=info msg=" - binding Unix socket" inherited=true socket=/var/snap/lxd/common/lxd/unix.socket
t=2021-05-06T16:16:11+0300 lvl=info msg="Initializing global database"
t=2021-05-06T16:16:30+0300 lvl=info msg="Firewall loaded driver \"xtables\""
t=2021-05-06T16:16:30+0300 lvl=info msg="Initializing storage pools"
t=2021-05-06T16:16:30+0300 lvl=info msg="Initializing daemon storage mounts"
t=2021-05-06T16:16:30+0300 lvl=info msg="Initializing networks"
t=2021-05-06T16:16:30+0300 lvl=info msg="Pruning leftover image files"
t=2021-05-06T16:16:30+0300 lvl=info msg="Done pruning leftover image files"
t=2021-05-06T16:16:30+0300 lvl=info msg="Loading daemon configuration"
t=2021-05-06T16:16:30+0300 lvl=info msg="Started seccomp handler" path=/var/snap/lxd/common/lxd/seccomp.socket
t=2021-05-06T16:16:30+0300 lvl=info msg="Pruning expired images"
t=2021-05-06T16:16:30+0300 lvl=info msg="Done pruning expired images"
t=2021-05-06T16:16:30+0300 lvl=info msg="Pruning expired instance backups"
t=2021-05-06T16:16:30+0300 lvl=info msg="Done pruning expired instance backups"
t=2021-05-06T16:16:30+0300 lvl=info msg="Expiring log files"
t=2021-05-06T16:16:31+0300 lvl=info msg="Done expiring log files"
t=2021-05-06T16:16:31+0300 lvl=info msg="Updating images"
t=2021-05-06T16:16:31+0300 lvl=info msg="Done updating images"
t=2021-05-06T16:16:31+0300 lvl=info msg="Updating instance types"
t=2021-05-06T16:16:31+0300 lvl=info msg="Done updating instance types"
>8

I've also tried to start lxd with "--debug" option but no relevant info was logged.

What extra info would be useful or what shall I try to help sort it out?

**Comment by stgraber** [comment]
Hmm, interesting, it's not something we've noticed so far, but with SSDs being so common, it's likely to have gone un-noticed :)

You could install the `fatrace` package and run `fatrace` while restarting LXD, this should show all file access so may help see what's different.

I can't think of any obvious difference though maybe it's something to do with the newer dqlite version.

**Comment by marius-nicolae** [comment]
Hm, I've never used fatrace and the captured events are weird and incomplete, I think. I've ran `fatrace -o /tmp/lxd_start1.txt -t` and then systemctl restart `snap.lxd.daemon`. The output is:
```
19:21:28.268543 snap(118832): RCO /snap/snapd/11588/usr/lib/snapd/info
19:21:28.268670 snap(118832): RO /snap/snapd/11588/usr/bin/snap
19:21:28.278070 snap(118832): RO /snap/snapd/11588/usr/lib/snapd/snapd
19:21:28.278070 snap(118832): R /snap/snapd/11588/usr/lib/snapd/snapd
19:21:28.278149 snap(118832): R /snap/snapd/11588/usr/lib/snapd/snapd
19:21:28.278149 snap(118832): R /snap/snapd/11588/usr/lib/snapd/snapd
19:21:28.278174 snap(118832): R /snap/snapd/11588/usr/lib/snapd/snapd
19:21:28.278186 snap(118832): R /snap/snapd/11588/usr/lib/snapd/snapd
19:21:28.278198 snap(118832): RC /snap/snapd/11588/usr/lib/snapd/snapd
19:21:28.279276 snap(118843): O /snap/snapd/11588/usr/lib/snapd/snap-seccomp
19:21:28.279349 snap(118843): R /snap/snapd/11588/usr/lib/snapd/snap-seccomp
19:21:28.279379 snap(118843): R /snap/snapd/11588/usr/lib/snapd/snap-seccomp
19:21:28.281007 snap-seccomp(118843): O /snap/snapd/11588/usr/lib/snapd/snap-seccomp
19:21:28.281050 snap-seccomp(118843): R /snap/snapd/11588/usr/lib/snapd/snap-seccomp
19:21:28.281050 snap-seccomp(118843): R /snap/snapd/11588/usr/lib/snapd/snap-seccomp
19:21:28.281086 snap-seccomp(118843): R /snap/snapd/11588/usr/lib/snapd/snap-seccomp
19:21:28.281110 snap-seccomp(118843): R /snap/snapd/11588/usr/lib/snapd/snap-seccomp
19:21:28.281122 snap-seccomp(118843): R /snap/snapd/11588/usr/lib/snapd/snap-seccomp
19:21:28.281133 snap-seccomp(118843): R /snap/snapd/11588/usr/lib/snapd/snap-seccomp
19:21:28.281146 snap-seccomp(118843): R /snap/snapd/11588/usr/lib/snapd/snap-seccomp
19:21:28.281146 snap-seccomp(118843): R /snap/snapd/11588/usr/lib/snapd/snap-seccomp
19:21:28.281169 snap-seccomp(118843): R /snap/snapd/11588/usr/lib/snapd/snap-seccomp
19:21:28.281181 snap-seccomp(118843): R /snap/snapd/11588/usr/lib/snapd/snap-seccomp
19:21:28.281193 snap-seccomp(118843): RC /snap/snapd/11588/usr/lib/snapd/snap-seccomp
19:21:28.282307 snap-seccomp(118843): C /snap/snapd/11588/usr/lib/snapd/snap-seccomp
19:21:28.282688 snap(118832): RCO /snap/lxd/20309/meta/snap.yaml
19:21:28.283427 snap(118832): RO /snap/lxd/20309/meta/hooks
19:21:28.283480 snap(118832): C /snap/lxd/20309/meta/hooks
19:21:28.284629 snap(118832): RO /snap/snapd/11588/usr/lib/snapd/snap-confine
19:21:28.284629 snap(118832): R /snap/snapd/11588/usr/lib/snapd/snap-confine
19:21:28.284706 snap(118832): R /snap/snapd/11588/usr/lib/snapd/snap-confine
19:21:28.285891 snap-confine(118832): C /snap/snapd/11588/usr/bin/snap
19:21:28.287547 snap-confine(118832): CO /snap/snapd/11588/usr/lib/snapd
19:21:28.287547 snap-confine(118832): O /snap/snapd/11588/usr/lib/snapd/snap-update-ns
19:21:28.287547 snap-confine(118832): O /snap/snapd/11588/usr/lib/snapd/snap-discard-ns
19:21:28.289011 snap-confine(118832): C /snap/snapd/11588/usr/lib/snapd/snap-discard-ns
19:21:28.289011 snap-confine(118832): C /snap/snapd/11588/usr/lib/snapd/snap-update-ns
19:21:28.289350 snap-exec(118832): C /snap/snapd/11588/usr/lib/snapd/snap-confine
19:21:31.336181 snap(118891): O /snap/snapd/11588/usr/lib/snapd/info
19:21:31.336227 snap(118891): RC /snap/snapd/11588/usr/lib/snapd/info
19:21:31.336247 snap(118891): O /snap/snapd/11588/usr/bin/snap
19:21:31.336266 snap(118891): R /snap/snapd/11588/usr/bin/snap
19:21:31.336284 snap(118891): R /snap/snapd/11588/usr/bin/snap
19:21:31.345874 snap(118891): RO /snap/snapd/11588/usr/lib/snapd/snapd
19:21:31.345874 snap(118891): R /snap/snapd/11588/usr/lib/snapd/snapd
19:21:31.345874 snap(118891): R /snap/snapd/11588/usr/lib/snapd/snapd
19:21:31.345952 snap(118891): R /snap/snapd/11588/usr/lib/snapd/snapd
19:21:31.345952 snap(118891): R /snap/snapd/11588/usr/lib/snapd/snapd
19:21:31.345975 snap(118891): R /snap/snapd/11588/usr/lib/snapd/snapd
19:21:31.345987 snap(118891): RC /snap/snapd/11588/usr/lib/snapd/snapd
19:21:31.347590 snap(118912): O /snap/snapd/11588/usr/lib/snapd/snap-seccomp
19:21:31.347667 snap(118912): R /snap/snapd/11588/usr/lib/snapd/snap-seccomp
19:21:31.347706 snap(118912): R /snap/snapd/11588/usr/lib/snapd/snap-seccomp
19:21:31.347730 snap(118912): R /snap/snapd/11588/usr/lib/snapd/snap-seccomp
19:21:31.349439 snap-seccomp(118912): RO /snap/snapd/11588/usr/lib/snapd/snap-seccomp
19:21:31.349439 snap-seccomp(118912): R /snap/snapd/11588/usr/lib/snapd/snap-seccomp
19:21:31.349439 snap-seccomp(118912): R /snap/snapd/11588/usr/lib/snapd/snap-seccomp
19:21:31.349535 snap-seccomp(118912): R /snap/snapd/11588/usr/lib/snapd/snap-seccomp
19:21:31.349550 snap-seccomp(118912): R /snap/snapd/11588/usr/lib/snapd/snap-seccomp
19:21:31.349565 snap-seccomp(118912): R /snap/snapd/11588/usr/lib/snapd/snap-seccomp
19:21:31.349580 snap-seccomp(118912): R /snap/snapd/11588/usr/lib/snapd/snap-seccomp
19:21:31.349580 snap-seccomp(118912): R /snap/snapd/11588/usr/lib/snapd/snap-seccomp
19:21:31.349609 snap-seccomp(118912): RC /snap/snapd/11588/usr/lib/snapd/snap-seccomp
19:21:31.350673 snap-seccomp(118912): C /snap/snapd/11588/usr/lib/snapd/snap-seccomp
19:21:31.351015 snap(118891): RCO /snap/lxd/20309/meta/snap.yaml
19:21:31.351575 snap(118891): RO /snap/lxd/20309/meta/hooks
19:21:31.351575 snap(118891): C /snap/lxd/20309/meta/hooks
19:21:31.353142 snap(118891): O /snap/snapd/11588/usr/lib/snapd/snap-confine
19:21:31.353191 snap(118891): R /snap/snapd/11588/usr/lib/snapd/snap-confine
19:21:31.353191 snap(118891): R /snap/snapd/11588/usr/lib/snapd/snap-confine
19:21:31.353191 snap(118891): R /snap/snapd/11588/usr/lib/snapd/snap-confine
19:21:31.354194 snap-confine(118891): C /snap/snapd/11588/usr/bin/snap
19:21:31.355431 snap-confine(118891): CO /snap/snapd/11588/usr/lib/snapd
19:21:31.355431 snap-confine(118891): O /snap/snapd/11588/usr/lib/snapd/snap-update-ns
19:21:31.355431 snap-confine(118891): O /snap/snapd/11588/usr/lib/snapd/snap-discard-ns
19:21:31.356935 snap-confine(118891): C /snap/snapd/11588/usr/lib/snapd/snap-discard-ns
19:21:31.356935 snap-confine(118891): C /snap/snapd/11588/usr/lib/snapd/snap-update-ns
19:21:31.357156 snap-exec(118891): C /snap/snapd/11588/usr/lib/snapd/snap-confine
19:21:31.383136 snap(118932): RCO /snap/snapd/11588/usr/lib/snapd/info
19:21:31.383136 snap(118932): O /snap/snapd/11588/usr/bin/snap
19:21:31.383233 snap(118932): R /snap/snapd/11588/usr/bin/snap
19:21:31.400002 snap(118932): C /snap/snapd/11588/usr/bin/snap
19:22:19.127729 snapd(1924): O /snap/snapd/11588/meta/snap.yaml
19:22:19.127794 snapd(1924): R /snap/snapd/11588/meta/snap.yaml
19:22:19.127826 snapd(1924): C /snap/snapd/11588/meta/snap.yaml 
```
However, the disk activity, at that restart took approx 2 minutes and 31 seconds.

Shall I run fatrace otherwise? Also, according to https://github.com/martinpitt/fatrace/issues/7 there seem to be some problems of fatrace with zfs.

**Comment by marius-nicolae** [comment]
Since it's a fresh Ubuntu install, the machine is mostly idle. The disk activity happens only when restarting the lxd snap service.

When running iotop, there are short moments when `lxd --logfile /var/snap/lxd/common/lxd/logs/lxd.log --group lxd` process gets to the top, as in the screenshot bellow
![image](https://user-images.githubusercontent.com/28258950/117358987-7fbcd380-aebf-11eb-8857-e244f4cbeecc.png)

but most of the time the `txg_sync` zfs utility process gets to the top, displaying a row such as 
```
520 be/4 root        0.00 B/s    0.00 B/s  0.00 % 21.68 % [txg_sync]
```

Also, the same disk activity is seen when running a mere `lxd --debug --group lxd`. In this case, also the following error is displayed:
```
EROR[05-06|23:19:38] Failed to start the daemon: LXD is already running
Error: LXD is already running
```
even though before starting it no lxd process was present (the service was stopped with `systemctl stop snap.lxd.daemon.service`).

Is there a way of running the lxd process in the foreground, with full debug messages (traces) activated, maybe without snap? Hopefully, we might get some meaningful info.

**Comment by stgraber** [comment]
- systemctl stop snap.lxd.daemon snap.lxd.daemon.unix.socket
 - lxd --debug --group lxd

**Comment by marius-nicolae** [comment]
Here is the output
```
t=2021-05-07T00:13:57+0300 lvl=dbug msg="Connecting to a local LXD over a Unix socket" 
t=2021-05-07T00:13:57+0300 lvl=dbug msg="Sending request to LXD" etag= method=GET url=http://unix.socket/1.0
t=2021-05-07T00:13:57+0300 lvl=info msg="LXD 4.0.6 is starting in normal mode" path=/var/snap/lxd/common/lxd
t=2021-05-07T00:13:57+0300 lvl=info msg="Kernel uid/gid map:" 
t=2021-05-07T00:13:57+0300 lvl=info msg=" - u 0 0 4294967295" 
t=2021-05-07T00:13:57+0300 lvl=info msg=" - g 0 0 4294967295" 
t=2021-05-07T00:13:57+0300 lvl=info msg="Configured LXD uid/gid map:" 
t=2021-05-07T00:13:57+0300 lvl=info msg=" - u 0 1000000 1000000000" 
t=2021-05-07T00:13:57+0300 lvl=info msg=" - g 0 1000000 1000000000" 
t=2021-05-07T00:13:57+0300 lvl=info msg="Kernel features:" 
t=2021-05-07T00:13:57+0300 lvl=info msg=" - closing multiple file descriptors efficiently: no" 
t=2021-05-07T00:13:57+0300 lvl=info msg=" - netnsid-based network retrieval: yes" 
t=2021-05-07T00:13:57+0300 lvl=info msg=" - pidfds: yes" 
t=2021-05-07T00:13:57+0300 lvl=info msg=" - uevent injection: yes" 
t=2021-05-07T00:13:57+0300 lvl=info msg=" - seccomp listener: yes" 
t=2021-05-07T00:13:57+0300 lvl=info msg=" - seccomp listener continue syscalls: yes" 
t=2021-05-07T00:13:57+0300 lvl=info msg=" - seccomp listener add file descriptors: no" 
t=2021-05-07T00:13:57+0300 lvl=info msg=" - attach to namespaces via pidfds: no" 
t=2021-05-07T00:13:57+0300 lvl=info msg=" - safe native terminal allocation : yes" 
t=2021-05-07T00:13:57+0300 lvl=info msg=" - unprivileged file capabilities: yes" 
t=2021-05-07T00:13:57+0300 lvl=info msg=" - cgroup layout: hybrid" 
t=2021-05-07T00:13:57+0300 lvl=warn msg=" - Couldn't find the CGroup blkio.weight, disk priority will be ignored" 
t=2021-05-07T00:13:57+0300 lvl=warn msg=" - Couldn't find the CGroup memory swap accounting, swap limits will be ignored" 
t=2021-05-07T00:13:57+0300 lvl=info msg=" - shiftfs support: yes" 
t=2021-05-07T00:13:57+0300 lvl=info msg="Initializing local database" 
t=2021-05-07T00:13:57+0300 lvl=dbug msg="Initializing database gateway" 
t=2021-05-07T00:13:57+0300 lvl=dbug msg="Start database node" address= id=1 role=voter
t=2021-05-07T00:13:57+0300 lvl=dbug msg="Connecting to a local LXD over a Unix socket" 
t=2021-05-07T00:13:57+0300 lvl=dbug msg="Sending request to LXD" etag= method=GET url=http://unix.socket/1.0
t=2021-05-07T00:13:57+0300 lvl=dbug msg="Detected stale unix socket, deleting" 
t=2021-05-07T00:13:57+0300 lvl=info msg="Starting /dev/lxd handler:" 
t=2021-05-07T00:13:57+0300 lvl=info msg=" - binding devlxd socket" socket=/var/snap/lxd/common/lxd/devlxd/sock
t=2021-05-07T00:13:57+0300 lvl=info msg="REST API daemon:" 
t=2021-05-07T00:13:57+0300 lvl=info msg=" - binding Unix socket" socket=/var/snap/lxd/common/lxd/unix.socket
t=2021-05-07T00:13:57+0300 lvl=info msg="Initializing global database" 
t=2021-05-07T00:13:57+0300 lvl=dbug msg="Dqlite: attempt 0: server 1: connected" 
t=2021-05-07T00:14:43+0300 lvl=dbug msg="Firewall xtables detected iptables is in use" 
t=2021-05-07T00:14:43+0300 lvl=info msg="Firewall loaded driver \"xtables\"" 
t=2021-05-07T00:14:43+0300 lvl=info msg="Initializing storage pools" 
t=2021-05-07T00:14:43+0300 lvl=dbug msg="Initializing and checking storage pool \"default\"" 
t=2021-05-07T00:14:43+0300 lvl=dbug msg="Mount started" driver=zfs pool=default
t=2021-05-07T00:14:43+0300 lvl=dbug msg="Mount finished" driver=zfs pool=default
t=2021-05-07T00:14:44+0300 lvl=info msg="Initializing daemon storage mounts" 
t=2021-05-07T00:14:44+0300 lvl=info msg="Initializing networks" 
t=2021-05-07T00:14:44+0300 lvl=dbug msg=Start driver=bridge network=lxdbr0
t=2021-05-07T00:14:44+0300 lvl=dbug msg="Setting up network" driver=bridge network=lxdbr0
t=2021-05-07T00:14:44+0300 lvl=dbug msg="Stable MAC generated" driver=bridge hwAddr=00:16:3e:88:2f:a6 network=lxdbr0 seed=58b579a3c700203b038dcc696ed1ac3d3de7a85a1aa45ffcd77842ab101c6b80.0.1
t=2021-05-07T00:14:44+0300 lvl=dbug msg="Clearing firewall" driver=bridge network=lxdbr0
t=2021-05-07T00:14:44+0300 lvl=dbug msg="Setting up firewall" driver=bridge network=lxdbr0
t=2021-05-07T00:14:44+0300 lvl=dbug msg="New task Operation: b9c7a2a6-5ab2-4024-9375-a8fc4a8321b5" 
t=2021-05-07T00:14:44+0300 lvl=info msg="Pruning leftover image files" 
t=2021-05-07T00:14:44+0300 lvl=dbug msg="Started task operation: b9c7a2a6-5ab2-4024-9375-a8fc4a8321b5" 
t=2021-05-07T00:14:44+0300 lvl=info msg="Done pruning leftover image files" 
t=2021-05-07T00:14:44+0300 lvl=info msg="Loading daemon configuration" 
t=2021-05-07T00:14:44+0300 lvl=dbug msg="Success for task operation: b9c7a2a6-5ab2-4024-9375-a8fc4a8321b5" 
t=2021-05-07T00:14:44+0300 lvl=dbug msg="Initialized inotify with file descriptor 26" 
t=2021-05-07T00:14:44+0300 lvl=info msg="Started seccomp handler" path=/var/snap/lxd/common/lxd/seccomp.socket
t=2021-05-07T00:14:44+0300 lvl=dbug msg="New task Operation: 7734e74d-ef31-4571-8619-2af0a71c7773" 
t=2021-05-07T00:14:44+0300 lvl=info msg="Pruning expired images" 
t=2021-05-07T00:14:44+0300 lvl=dbug msg="Started task operation: 7734e74d-ef31-4571-8619-2af0a71c7773" 
t=2021-05-07T00:14:44+0300 lvl=info msg="Done pruning expired images" 
t=2021-05-07T00:14:44+0300 lvl=dbug msg="New task Operation: fe3633e0-b038-46b2-a922-dce2f24e9260" 
t=2021-05-07T00:14:44+0300 lvl=dbug msg="Success for task operation: 7734e74d-ef31-4571-8619-2af0a71c7773" 
t=2021-05-07T00:14:44+0300 lvl=info msg="Pruning expired instance backups" 
t=2021-05-07T00:14:44+0300 lvl=dbug msg="Started task operation: fe3633e0-b038-46b2-a922-dce2f24e9260" 
t=2021-05-07T00:14:44+0300 lvl=info msg="Done pruning expired instance backups" 
t=2021-05-07T00:14:44+0300 lvl=dbug msg="Success for task operation: fe3633e0-b038-46b2-a922-dce2f24e9260" 
t=2021-05-07T00:14:44+0300 lvl=dbug msg="New task Operation: a8ba88db-5327-473b-a55b-453012e780c5" 
t=2021-05-07T00:14:44+0300 lvl=info msg="Updating instance types" 
t=2021-05-07T00:14:44+0300 lvl=dbug msg="Started task operation: a8ba88db-5327-473b-a55b-453012e780c5" 
t=2021-05-07T00:14:44+0300 lvl=info msg="Done updating instance types" 
t=2021-05-07T00:14:44+0300 lvl=dbug msg="New task Operation: cc2f00a6-6370-46af-b190-644812595a43" 
t=2021-05-07T00:14:44+0300 lvl=info msg="Expiring log files" 
t=2021-05-07T00:14:44+0300 lvl=dbug msg="Started task operation: cc2f00a6-6370-46af-b190-644812595a43" 
t=2021-05-07T00:14:44+0300 lvl=info msg="Done expiring log files" 
t=2021-05-07T00:14:44+0300 lvl=dbug msg="Success for task operation: cc2f00a6-6370-46af-b190-644812595a43" 
t=2021-05-07T00:14:47+0300 lvl=dbug msg="Success for task operation: a8ba88db-5327-473b-a55b-453012e780c5" 
t=2021-05-07T00:16:34+0300 lvl=info msg="Received 'interrupt signal', exiting" 
t=2021-05-07T00:16:34+0300 lvl=dbug msg="Cancel ongoing or future gRPC connection attempts" 
t=2021-05-07T00:16:34+0300 lvl=info msg="Starting shutdown sequence" 
t=2021-05-07T00:16:34+0300 lvl=info msg="Closing the database" 
t=2021-05-07T00:16:34+0300 lvl=info msg="Stop database gateway" 
t=2021-05-07T00:16:34+0300 lvl=info msg="Stopping REST API handler:" 
t=2021-05-07T00:16:34+0300 lvl=info msg=" - closing socket" socket=/var/snap/lxd/common/lxd/unix.socket
t=2021-05-07T00:16:34+0300 lvl=info msg="Stopping /dev/lxd handler:" 
t=2021-05-07T00:16:34+0300 lvl=info msg=" - closing socket" socket=/var/snap/lxd/common/lxd/devlxd/sock
t=2021-05-07T00:16:34+0300 lvl=info msg="Unmounting temporary filesystems" 
t=2021-05-07T00:16:34+0300 lvl=info msg="Done unmounting temporary filesystems"
```

The last logged line while starting is `t=2021-05-07T00:14:47+0300 lvl=dbug msg="Success for task operation: a8ba88db-5327-473b-a55b-453012e780c5" ` but the disk activity continued for another minute, or so (might be the zfs cache flushing).

**Comment by stgraber** [comment]
Yeah, one reason could be that our database is a lot less crap now than it was in 3.0, I wouldn't be surprised if back in 3.0 we weren't flushing things to disk correctly. Leading to less disk access, sure, but also, you lose power and your data is gone...

**Comment by marius-nicolae** [comment]
Wouldn't that data be flushed at a restart? I imagine that when the lxd process gets destroyed all the file descriptors are closed, triggering a disk flash (by kernel).

**Comment by stgraber** [comment]
Yes, the issue tends to be with things like power loss.

**Comment by stgraber** [comment]
Yeah, the main difference being that with dqlite flushing for every transaction, you've got a lot more small disk writes with the safe I/O behavior vs just a few big ones with the unsafe one.

**Comment by marius-nicolae** [comment]
For me, it's hard to imagine how many write transactions a clean LXD install, without any container, would need just to start, so that it keeps a hard drive busy for around two and half minutes. In this case, I would rather see the database as some kind of a more advanced configuration file, which should translate in some reads and much fewer writes.

Anyway, is there any possibility to disable those fsyncs? Usually, the servers have UPSes so the power outage wouldn't cause so much of an impact. 

If there are no other more "official" approaches, as a workaround, I'm thinking to play with [eatmydata](http://manpages.ubuntu.com/manpages/trusty/man1/eatmydata.1.html) and I'm not sure if it will work also with snaps.

**Comment by stgraber** [comment]
@marius-nicolae just to confirm that this is still a problem before I nag @brauner again to look into it?
